conformance/utils/kubernetes/apply.go

```diff
@@ -41,45 +42,95 @@ import (
+// returns the ports used by the listeners if they came from validPorts.
+func (a *Applier) prepareGateway(t *testing.T, uObj *unstructured.Unstructured, gatewayClassName string) {
+	name := types.NamespacedName{
+		Namespace: uObj.GetNamespace(),
+		Name:      uObj.GetName(),
+	}
+	a.freePortsFor(name)
+
 	err := unstructured.SetNestedField(uObj.Object, gatewayClassName, "spec", "gatewayClassName")
 	require.NoErrorf(t, err, "error setting `spec.gatewayClassName` on %s Gateway resource", uObj.GetName())
 
-	if len(validListenerPorts) > 0 {
-		listeners, _, err := unstructured.NestedSlice(uObj.Object, "spec", "listeners")
-		require.NoErrorf(t, err, "error getting `spec.listeners` on %s Gateway resource", uObj.GetName())
+	var allocatedPorts []v1alpha2.PortNumber
 
-		for i, uListener := range listeners {
-			require.Less(t, portIndex, len(validListenerPorts), "not enough unassigned valid ports for `spec.listeners[%d]` on %s Gateway resource", i, uObj.GetName())
+	if len(a.availablePorts) > 0 {
+		uListeners, _, err := unstructured.NestedSlice(uObj.Object, "spec", "listeners")
+		require.NoErrorf(t, err, "error getting `spec.listeners` on %s Gateway resource", uObj.GetName())
 
+		var listeners []interface{}
+		newPorts := map[int64]v1alpha2.PortNumber{}
+		for i, uListener := range uListeners {
 			listener, ok := uListener.(map[string]interface{})
 			require.Truef(t, ok, "unexpected type at `spec.listeners[%d]` on %s Gateway resource", i, uObj.GetName())
 
-			nextPort := validListenerPorts[portIndex]
-			err = unstructured.SetNestedField(listener, int64(nextPort), "port")
-			require.NoErrorf(t, err, "error setting `spec.listeners[%d].port` on %s Gateway resource", i, uObj.GetName())
+			port, _, portErr := unstructured.NestedInt64(listener, "port")
+			require.NoErrorf(t, portErr, "error getting `spec.listeners[%d].port` on %s Gateway resource", i, uObj.GetName())
+
+			// For each listener port either allocate a new port or use the port
+			// already allocated for this listener port
+			newPort, ok := newPorts[port]
+			if !ok {
+				var portIsValid bool
+				newPort, portIsValid = a.availablePorts.AssignAvailablePort()
+				require.True(t, portIsValid, "not enough unassigned valid ports for Gateway resource")
+				newPorts[port] = newPort
+			}
+
+			portErr = unstructured.SetNestedField(listener, int64(newPort), "port")
+			require.NoErrorf(t, portErr, "error setting `spec.listeners[%d].port` on %s Gateway resource", i, uObj.GetName())
 
-			portIndex++
-			listeners[i] = listener
+			listeners = append(listeners, listener)
+		}
+
+		for _, port := range newPorts {
+			allocatedPorts = append(allocatedPorts, port)
 		}
 
 		err = unstructured.SetNestedSlice(uObj.Object, listeners, "spec", "listeners")
 		require.NoErrorf(t, err, "error setting `spec.listeners` on %s Gateway resource", uObj.GetName())
 	}
 
-	return portIndex
+	a.assignedPorts[name] = allocatedPorts
```

Comment:
I'm not sure how easy this is, but it seems like it would be much safer to pull all the port logic out to a separate function and handle locking inside there. All these calls like this feel rather dangerous given the size of this function and that it's unclear if/when locking has occurred.

Reply:
The lock actually belongs in `prepareGateway` so I've put it there, which at least keeps it in this function. I'm not sure how to refactor to remove the port logic any further or whether it would help. The only thing `prepareGateway` does besides set the gateway class is iterating through the listeners to do this port logic.